edx_api/enrollments/__init__.py

```diff
@@ -22,6 +25,88 @@ def __init__(self, requester, base_url):
         """
         self.requester = requester
         self.base_url = base_url
+        self._course_details = {}
+
+    def _get_course_details(self, course_id):
+        """
+        Retrieves course details.
+
+        Args:
+            course_id (str): Course id.
+        """
+        if course_id in self._course_details:
+            return self._course_details[course_id]
+
+        req_url = urljoin(self.base_url, self.course_details_url.format(course_key=course_id))
+        resp = self.requester.get(req_url)
+        resp.raise_for_status()
+        resp_json = resp.json()
+        self._course_details[course_id] = resp_json
+        return resp_json
+
+    def _parse_enrollments(self, enrollments_list):
+        """
+        Parse enrollments response into :class:`Enrollments`.
+        """
+        def populate_course_details(enrollment):
+            """Populates course details if missing."""
+            enrollment.pop('previous', None)
+            enrollment.pop('next', None)
+            if enrollment.get('course_details'):
+                return enrollment
+            course_details = self._get_course_details(enrollment['course_id'])
+            course_details['course_start'] = course_details['start']
+            course_details['course_end'] = course_details['end']
+            enrollment['course_details'] = course_details
+            return enrollment
```

Comment:
All of this parsing seems an unnecessary byproduct of trying to use the existing `Enrollments` model that also causes the issue I described below which discards all the data.

That `Enrollments` model was designed around the results which come from the other API endpoint, and internally handles the parsing of that data.

We shouldn't need to coax this API's results into another form. It's adding a lot of extra code here, and already causing some issues with the results it returns.

Also we're now having to call into another API possibly hundreds of times from this one request to fill in data that the other API returns implicitly.

We should really find some other solution, possibly creating a new model to parse the results from this endpoint, or something other than this.

Reply:
I agree that it would be better to create a new class. I'll do that.